docroot/modules/custom/civictheme_migrate/src/ContentComponentFactory.php

```diff
@@ -0,0 +1,228 @@
+<?php
+
+namespace Drupal\civictheme_migrate;
+
+use Drupal\Component\Utility\NestedArray;
+use Drupal\Core\Database\Connection;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\civictheme_migrate\Utils\CivicthemeTrait;
+use Drupal\migrate\Plugin\MigratePluginManager;
+use Drupal\paragraphs\ParagraphInterface;
+use Symfony\Component\Serializer\NameConverter\CamelCaseToSnakeCaseNameConverter;
+
+/**
+ * Content Component Factory.
+ *
+ * @package Drupal\civictheme_migrate
+ */
+class ContentComponentFactory {
+  use CivicthemeTrait;
+
+  /**
+   * The MigratePluginManager instance.
+   *
+   * @var \Drupal\migrate\Plugin\MigratePluginManagerInterface
+   */
+  protected $migratePluginManager;
+
+  /**
+   * Entity Type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * ContentComponentFactory constructor.
+   *
+   * @param \Drupal\migrate\Plugin\MigratePluginManager $migrate_plugin_manager
+   *   Migrate plugin manager.
+   * @param \Drupal\Core\Database\Connection $database
+   *   The database connection.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   Entity type manager.
+   */
+  public function __construct(MigratePluginManager $migrate_plugin_manager, Connection $database, EntityTypeManagerInterface $entity_type_manager) {
+    $this->migratePluginManager = $migrate_plugin_manager;
+    $this->database = $database;
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * Create components.
+   *
+   * @param mixed $items
+   *   The list of items to create.
+   * @param array $context
+   *   The migration context array with 4 keys:
+   *    - configuration: the configuration of the current field.
+   *    - migration: the current migrate executable.
+   *    - row: the current row.
+   *    - destination_property: destination property currently worked on.
+   *
+   * @return array|null
+   *   The components.
+   */
+  public function createComponents($items, array &$context) {
+    if (!is_array($items)) {
+      return NULL;
+    }
+
+    $components = [];
+    foreach ($items as $item) {
+      foreach ($item as $item_type => $item_data) {
+        $component = $this->createComponent($item_type, $item_data, $context);
+        if ($component instanceof ParagraphInterface) {
```

Comment:
please add a comment here in code to explain in which cases `$component` will be a `ParagraphInterface` instance (would it get pre-transformed by the migration?)